_posts/henry_zhao/2019-03-06-Implement-Stardust.md

```diff
@@ -0,0 +1,287 @@
+        );
+        t1.check(e);
+      } else {
+        circles2.data([]);
+        circles2.render();
+      }
+    };
+    this._canvasNode.onmousedown = e => {
+      let bounds = this._canvasNode.getBoundingClientRect();
+      var x = e.clientX - bounds.left;
+      var y = e.clientY - bounds.top;
+      var pos = { x: x, y: y };
+      window.open().document.write(JSON.stringify(pos));
+    };
+  }
+  addTooltips() {}
+  // if data looks like this: data = [{ x: 16, y: 1 }, { x: 2, y: 3 }, {x: 4, y: 7}] and width is 20;
+  // chartData will look like this: chartData = [{p1: [], p2: []}, {p1: [], p2: []}, {p1:[], p2: []}]
+  setChartData(data, rowCount, colCount, height, width) {
+    console.log('rect data size', data.filter(d => d.show).length);
+
+    let chartData = data.filter(d => d.show).map(d => ({
+      p1: [
+        (width / rowCount + 1) * d.x + 20,
+        (height / colCount + 1) * d.y + 20,
+      ],
+      p2: [
+        (width / rowCount + 1) * d.x + 20 + width / rowCount,
+        (height / colCount + 1) * d.y + 20 + height / colCount,
+      ],
+    }));
+    return chartData;
+  }
+  rendered() {
+    var rectData = [];
+    for (var i = 0; i < 250; i++) {
+      for (var j = 0; j < 100; j++) {
+        var random_boolean = Math.random() >= 0.6;
+        rectData.push({ x: i, y: j, show: random_boolean });
+      }
+    }
+    var circleData = [];
+    for (var i = 0; i < 250; i++) {
+      for (var j = 0; j < 100; j++) {
+        var random_boolean = Math.random() >= 0.6;
+        circleData.push({ x: i, y: j, show: random_boolean });
+      }
+    }
+    circleData = circleData.filter(d => d.show);
+    this._marks.rects.attr('p1', d => d['p1']);
+    this._marks.rects.attr('p2', d => d['p2']);
+    this._marks.rects.attr('color', d => [0, 254, 127, 1]);
+    this._marks.rects.data(this.setChartData(rectData, 250, 100, 600, 1500));
+    this._marks.rects.render();
+
+    this._marks.circles.attr('center', d => [7 * d.x + 23, 7 * d.y + 23]);
+    this._marks.circles.attr('radius', 2);
+    this._marks.circles.attr('color', [0, 0, 0, 1]);
+    this._marks.circles.data(circleData);
+    this._marks.circles.render();
+    this._platform.beginPicking(
+      this._canvasNode.width,
+      this._canvasNode.height
+    );
+    this._marks.rects.render();
+    this._marks.circles.render();
+    this._platform.endPicking();
+  }
+}
+```
+
+### Conclusion
+
+Although my implementation doesn't cover many real scenarios, there is no doubt that stardust has a better rendering performance than d3. A fully replacement of d3 will never happen in a real world due to the all the powerful components d3 provided and the advantages of svg in event handling. D3 are also a good complementary for Stardust(WebGL). A mix of the two in applications is recommended by the creator of Stardust: d3 being used to render a scatterplot’s axes and handle interactions such as range selections, but Stardust used to render and animate the points.
```

Comment:
If you rephrase a bit your second sentence to actually introduce your last point earlier. For example:
`Nevertheless, Stardust doesn't aim at replacing d3, but more at improving performance in rendering and animating points. The creator of Stardust actually recommends....`